From 137ca52faf24e360a147c9941f5ee98445a2543f Mon Sep 17 00:00:00 2001
From: Piyush Mehta <piyush.mehta@xilinx.com>
Date: Thu, 23 Jan 2020 19:51:12 +0530
Subject: [PATCH 1023/1852] usb: dwc3: gadget: get irq number for hibernation
 interrupts

commit ce5078825395addecc6628c4f58ef3d72d15f255 from
https://github.com/Xilinx/linux-xlnx.git

This patch modifies the code to get irq number for
usb hibernation specific interrupts.

Signed-off-by: Piyush Mehta <piyush.mehta@xilinx.com>
Signed-off-by: Anurag Kumar Vulisha <anurag.kumar.vulisha@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/usb/dwc3/gadget.c | 37 +++++++++++--------------------------
 1 file changed, 11 insertions(+), 26 deletions(-)

diff --git a/drivers/usb/dwc3/gadget.c b/drivers/usb/dwc3/gadget.c
index 416ed9a35f7b..e2ecf714f5a7 100644
--- a/drivers/usb/dwc3/gadget.c
+++ b/drivers/usb/dwc3/gadget.c
@@ -3450,7 +3450,7 @@ static irqreturn_t dwc3_interrupt(int irq, void *_evt)
 static int dwc3_gadget_get_irq(struct dwc3 *dwc)
 {
 	struct platform_device *dwc3_pdev = to_platform_device(dwc->dev);
-	int irq;
+	int irq, irq_hiber;
 
 	irq = platform_get_irq_byname_optional(dwc3_pdev, "peripheral");
 	if (irq > 0)
@@ -3473,24 +3473,20 @@ static int dwc3_gadget_get_irq(struct dwc3 *dwc)
 	if (irq == -EPROBE_DEFER)
 		goto out;
 
+out:
 	/* look for wakeup interrupt if hibernation is supported */
 	if (dwc->has_hibernation) {
-		irq = platform_get_irq(dwc3_pdev, 2);
-		if (irq > 0)
-			dwc->irq_wakeup = irq;
 
-		if (irq == -EPROBE_DEFER)
-			goto out;
+		irq_hiber = platform_get_irq_byname(dwc3_pdev, "hiber");
+		if (irq_hiber > 0) {
+			dwc->irq_wakeup = irq_hiber;
+		} else {
+			irq_hiber = platform_get_irq(dwc3_pdev, 2);
+			if (irq_hiber > 0)
+				dwc->irq_wakeup = irq_hiber;
+		}
 	}
 
-	if (irq <= 0) {
-		if (irq != -EPROBE_DEFER)
-			dev_err(dwc->dev, "missing peripheral IRQ\n");
-
-		if (!irq)
-			irq = -EINVAL;
-	}
-out:
 	return irq;
 }
 
@@ -3511,18 +3507,7 @@ int dwc3_gadget_init(struct dwc3 *dwc)
 		goto err0;
 	}
 
-	if (dwc->dr_mode == USB_DR_MODE_OTG) {
-		struct usb_phy *phy;
-
-		/* Switch otg to peripheral mode */
-		phy = usb_get_phy(USB_PHY_TYPE_USB3);
-		if (!IS_ERR(phy)) {
-			if (phy && phy->otg)
-				otg_set_peripheral(phy->otg,
-						(struct usb_gadget *)1);
-			usb_put_phy(phy);
-		}
-	}
+	dwc->irq_gadget = irq;
 
 	dwc->ep0_trb = dma_alloc_coherent(dwc->sysdev,
 					  sizeof(*dwc->ep0_trb) * 2,
-- 
2.31.1

